core/src/main/scala/cats/syntax/monadError.scala

```diff
@@ -16,6 +16,9 @@ final class MonadErrorOps[F[_], E, A](val fa: F[A]) extends AnyVal {
   def ensureOr(error: A => E)(predicate: A => Boolean)(implicit F: MonadError[F, E]): F[A] =
     F.ensureOr(fa)(error)(predicate)
 
+  def reject(pf: PartialFunction[A, E])(implicit F: MonadError[F, E]): F[A] =
+    F.flatMap(fa)(a => if (pf.isDefinedAt(a)) F.raiseError(pf(a)) else F.pure(a))
```

Comment:
There is [this trick](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/Foldable.scala#L210) that avoids calling `isDefinedAt` and then `apply` of the `PartialFunction`. Also, instead of `F.pure(a)` shall we just return `fa`?

Reply:
I'm thinking of doing `F.flatMap(fa)(a => pf.andThen(F.raiseError[A]).applyOrElse(a, _ => fa))`, though that does create a new "AndThen" object. Is avoiding the allocation important?

Reply:
that might be a good balance between readability and performance.